### ISSUE TITLE
Add fastText to cltkv1 (was: CLTK 2.0 fastText inclusion proposal)

### ISSUE THREAD

**Issue by todd-cook** [cltk/cltkv1#10]
fastText is a new library to create vector models of words, it has been developed and released by Facebook AI team.
https://github.com/facebookresearch/fastText
https://fasttext.cc/docs/en/aligned-vectors.html

fastText representations come in two forms, the standard text vector file format (akin to Glove), and a large binary model format which must be interacted with through the fastText API.
The text vector file of embeddings behaves just like all others (word2vec, glove, etc).

The binary models of fastText provide the ability to produce embeddings for out-of-vocabulary words; they can do this because they track sub-word level tokens. This is a game-changing feature which would be very desirable for CLTK, since many Classical texts have hapax legomenon, which wreak havoc on modeling because of dropout. The values of the projected embeddings are dependent on the neighboring subwords, but generally they provide an important hint to hapax's implicit meaning.

Costs we should consider: 
* the models are very large!
* the fastText native API is C++ (with Python wrappers) 
The native code require (g++-4.7.2 or newer) or (clang-3.3 or newer) and pybind11, and although it is excellent and will probably be well supported, it could be a concern in the future.

Facebook has released pre-trained word vectors and models in 157 Languages, trained on Common Crawl and Wikipedia : https://fasttext.cc/docs/en/crawl-vectors.html

e.g. for Latin, there is a binary model:
cc.la.300.bin.gz 3.2GB compressed (4.7 GB uncompressed) 
cc.la.300.vec.gz  (616 MB compressed), (2.3 GB uncompressed)

CLTK base models would want to train their own fastText representations--I have, and it's not too hard, but it does require quite a bit of memory and swap.

I think this is a game changing feature that we should consider adopting into the CLTK 2.0 paradigm. 
Thoughts? Comments? Questions?

**Comment by kylepjohnson** [comment]
> The binary models of fastText provide the ability to produce embeddings for out-of-vocabulary words; they can do this because they track sub-word level tokens. This is a game-changing feature which would be very desirable for CLTK, since many Classical texts have hapax legomenon, which wreak havoc on modeling because of dropout. The values of the projected embeddings are dependent on the neighboring subwords, but generally they provide an important hint to hapax's implicit meaning.

This makes tons of sense, thanks for the explanation. Like word2vec, are can with skip-grams instead of n-grams? This might help with non-fixed word order languages.

Making note of in-scope languages, from a quick skim [of available models](https://fasttext.cc/docs/en/crawl-vectors.html):
* Latin
* Sanskrit

@todd-cook sounds like you have a plan to train more of our own, right? An idea of roughly how many words are required for a tractable model?

> I think this is a game changing feature that we should consider adopting into the CLTK 2.0 paradigm.
Thoughts? Comments? Questions?

I am all for this. For some reason, from your recent post, I was thinking that fasttext would be an alternative to stanfordnlp (cc @free-variation ) however these can stand independently and not interfere with one another, right? I think this would be wise, as we continue to feel out how to implement them.

> the fastText native API is C++ (with Python wrappers)
The native code require (g++-4.7.2 or newer) or (clang-3.3 or newer) and pybind11, and although it is excellent and will probably be well supported, it could be a concern in the future.

My biggest concern is that we make installation as easy on our users as possible. It isn't that long ago -- I can still remember my own project-killing installation woes faced as a semi-literate programmer. Also, of course, we should have non-breaking fallbacks for users who do not have, and do not want, the fasttext dependencies.

I'm all ears, tell me if you want more inputs, otherwise I will await your first code draft. Also, if you think it would help communication, merge my PR #8 and build off that.

**Comment by free-variation** [comment]
Hefty +1 from Brooklyn on this one :)  I had to induce my own word2vec vectors when training the POS/morpho tagger for OE using stanfordnlp.  I did so by hand, rather than use gensim, because I wanted to get it done quickly using a GPU.  It worked, but the hapax/OOV problem bugs me.

(An interesting observation: the most recent Keras-based POS tagger I have doesn't use vectors, just subword features, yet it performs very near stanfordnlp.  Something to reflect on.)

@kylepjohnson yes, stanfordnlp and FastText are complementary.  In fact, when one downloads the full set of vectors for stanfordnlp, their download script pulls the vectors from Facebook for many of the languages for which there are models.

**Comment by kylepjohnson** [comment]
@free-variation you want at this one?

What would be most appropriate with the current `Doc` and `Word` dataclasses? Would you add a new attribute to `Doc`, say `Doc.embeddings` and add a list of lists there, that is indexed to the word tokens? Or do a `Word.embedding`, with an embedding tied to the word?

Of course, it'd be ideal to have one general use interface for users to get any available word embedding from any "vendor". But if we do not currently foresee using any other embedding vendor, then let's tie to fasttext.

I'd like to use this as an opportunity to see if another can make use of the `Process` type, too. You can find patterns to start with at `cltkv1/tokenizers/word.py`.

**Comment by kylepjohnson** [comment]
I've done some work to include fastText. I'm not satisfied with either the Python library and/or my understanding of it.

### The Good

* It has embeddings for at least 7 in-scope languages: https://github.com/kylepjohnson/cltkv1/blob/add-embeddings/src/cltkv1/embeddings/fasttext_module.py#L26


### The Bad

* The biggest issue is that I get errors, I suspect to do with insufficient memory, that say my models (downloaded from the ft website) are in the incorrect format. One model consistently works on my personal machine (wikipedia for Latin), while the rest fail. I'll put example below.
* Installation of the base C++ is hit-or-miss for me. If it can truly be installed through a `pip install`, then that's great … otherwise we have to think hard about what a user will tolerate.
* Their model downloading fn did not work for me out of the box, so I wrote my own: https://github.com/facebookresearch/fastText/blob/02c61efaa6d60d6bb17e6341b790fa199dfb8c83/python/fasttext_module/fasttext/util/util.py#L183

Usage example that fails for me:

``` python
$ poetry run ipython
Python 3.7.5 (default, Oct 31 2019, 20:57:45) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.11.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import fasttext                                                                                                                               

In [2]: f = "/Users/kyle.p.johnson/cltk_data/lat/embeddings/fasttext/cc.la.300.bin"                                                                   

In [3]: model = fasttext.load_model(f)                                                                                                                

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-979350476212> in <module>
----> 1 model = fasttext.load_model(f)

~/.pyenv/versions/3.7.5/envs/cltk/lib/python3.7/site-packages/fasttext/FastText.py in load_model(path)
    348     """Load a model given a filepath and return a model object."""
    349     eprint("Warning : `load_model` does not return WordVectorModel or SupervisedModel any more, but a `FastText` object which is very similar.")
--> 350     return _FastText(model_path=path)
    351 
    352 

~/.pyenv/versions/3.7.5/envs/cltk/lib/python3.7/site-packages/fasttext/FastText.py in __init__(self, model_path, args)
     41         self.f = fasttext.fasttext()
     42         if model_path is not None:
---> 43             self.f.loadModel(model_path)
     44         self._words = None
     45         self._labels = None

ValueError: /Users/kyle.p.johnson/cltk_data/lat/embeddings/fasttext/cc.la.300.bin has wrong file format!
```

There is very little help online with this problem … am I possibly missing something obvious? Should I go to them for help? 

Have either of you run these [ft models through gensim](https://radimrehurek.com/gensim/models/fasttext.html)?

**Comment by free-variation** [comment]
Let me try this evening.  I have two thinkpads running different OSs so there may be enlightening differences.

**Comment by kylepjohnson** [comment]
Thanks, John.

I should have mentioned I did a little memory profiling ([with this lib](https://pypi.org/project/memory-profiler/)) but didn't get anything useful from it, as fasttext would complain about the model's formatting before the thing could get into memory. But honestly I was probably misunderstanding it.

If memory does seem to be the matter, let me know and I'll go to the bother of running on a bigger rig.

**Comment by todd-cook** [comment]
I'll check my past efforts, but from memory, in my experience:
* I got a particular version of the fasttext python library to load the files ( and I think this required the c++11 install). 
* I found that memory ballooned on OSX. I tried to run fasttext on moderate sized corpus and ran out of memory near the end of the process.
* I was able to successfully run fastext on ubuntu and generate the binary and text (wordvec format) files.

The binary models are huge and basically of a proprietary format, and I would prefer using the wordvec (text) format.
Since the majority of the languages we care about are discrete, I assumed that we could process all our corpora, produce a binary, and then distill out embeddings for all known words including hapax legomena, and then export this text (wordvec) embedding for CLTK use and distribution. I don't trust the lifespan guarantee of the fasttext bin model, and since the vocab space is largely known, why not produce a discrete embedding?

I do think there's room and potential for generating ELMO models for the classical languages, so binary isn't out of the question, and ELMO would provide value over the embeddings. But for most CLTK users, embeddings will be more accessible.

The question remains, what is a really good set of embedding settings to train on. I have a few, but I was hoping for the embedding benchmarking info before proceeding with this (@diyclassics ?)

**Comment by kylepjohnson** [comment]
> The binary models are huge and basically of a proprietary format, and I would prefer using the wordvec (text) format.

Todd, do you know whether Gensim can run the .vec models provided by fasttext? I thought I read this, but didn't try for myself.

>  I don't trust the lifespan guarantee of the fasttext bin model, and since the vocab space is largely known, why not produce a discrete embedding?

I agree in spirit, however until we have this finished, I want us to provide something in the short-term, too.

**Comment by todd-cook** [comment]
> Todd, do you know whether Gensim can run the .vec models provided by fasttext? I thought I read this, but didn't try for myself.
I think I was able to use Gensim; I think there's two ways to load a model and I could only get one of them to work. I'll check my code and try to provide a snippet.

**Comment by kylepjohnson** [comment]
> I think I was able to use Gensim; I think there's two ways to load a model and I could only get one of them to work. 

This would be great. I should have started with Gensim, probably, however I wanted to give the fasttext library a fair shake. You make a good argument, too, about using the full text (.vec) files. Unless they're causing memory errors on ordinary laptops, it makes sense to default to these.

**Comment by free-variation** [comment]
@kylepjohnson on a 24GB system running win10 and using the linux subsystem, the *manually downladed* cc.la.300.bin model loaded without errors.

Note that I had to `gunzip` the manually downloaded model.  Do you think something may have become corrupted in the downloading process?

I'll try now on a native linux install.

**Comment by kylepjohnson** [comment]
> Note that I had to gunzip the manually downloaded model

It's possible. I'll starting with all three, from a manual, and report back.

**Comment by free-variation** [comment]
On a 32GB Linux Mint (ubuntu) system also no issues -- and man, that 3GB file loaded in a couple of seconds.

**Comment by kylepjohnson** [comment]
Also, which models were you using? Did you try all 4 Latin ones?

**Comment by free-variation** [comment]
No, just the one you have in the trace above: `cc.la.300.bin`.

**Comment by kylepjohnson** [comment]
> On a 32GB Linux Mint (ubuntu) system also no issues -- and man, that 3GB file loaded in a couple of seconds.

Relieved it works somewhere!

Could you make it work on a mac, too?

**Comment by kylepjohnson** [comment]
- fresh download of all models through browser
- unzip through Mac window browser
- Clean build of the C++ app (`make`), then two .bin for Latin are good; the all three .vec fail (one English, two Latin).

John, any guesses why or advice about how to troubleshoot the .vec models, especially for memory problems?

``` shell
$ ./fasttext nn /Users/kyle.p.johnson/Downloads/wiki.la.vec 
libc++abi.dylib: terminating with uncaught exception of type std::invalid_argument: /Users/kyle.p.johnson/Downloads/wiki.la.vec has wrong file format!
Abort trap: 6

$ ./fasttext nn /Users/kyle.p.johnson/Downloads/wiki-news-300d-1M.vec 
libc++abi.dylib: terminating with uncaught exception of type std::invalid_argument: /Users/kyle.p.johnson/Downloads/wiki-news-300d-1M.vec has wrong file format!
Abort trap: 6

$ ./fasttext nn /Users/kyle.p.johnson/Downloads/cc.la.300.bin
Query word? amicitia
amicitiâ 0.640267
amicitiá 0.638511
Inimicitia 0.626936
amicitiaque 0.622871
Amicitia 0.60711
amicitias 0.594631
amicitiae 0.593753
inimicitia 0.572094
amicitiam 0.568586
amicitiarum 0.564544

$ ./fasttext nn /Users/kyle.p.johnson/Downloads/wiki.la/wiki.la.bin
Query word? amicitia 
amicitiam 0.862521
amicitiae 0.857107
amicitias 0.845426
inimicitia 0.799734
amicizia 0.695931
pudicitia 0.688895
inimicitiam 0.677555
amici 0.674497
inimicitiae 0.674045
inimicitias 0.670258
```

**Comment by todd-cook** [comment]
Here's the code I used to load vec files.
They were created by me, but the Gensim loading should work for the wiki.vec releases.


from gensim.models import KeyedVectors

def assess_word2vec(latin_words, embeddings_root, kvfile):
    print('assessing', kvfile)
    try:
        vecwords = set() 
        filepath = os.path.join( embeddings_root, kvfile)  
        embeddings_index = {}
        wv_from_bin = KeyedVectors.load_word2vec_format(filepath, binary=True) 
        for word, vector in zip(wv_from_bin.vocab, wv_from_bin.vectors):
            coefs = np.asarray(vector, dtype='float32')
            embeddings_index[word] = coefs
            vecwords.add(word.lower())
        print(len(vecwords.intersection(latin_words)), len(latin_words))
    except:
        print('Fail')


vecs="""latin.lib.fasttext.vec 
latin_library..2019.05.09.vec 
latin_library.2019.03.06.vec 
latin_library.2019.03.07.vec 
latin_library.2019.03.08.vec
latin_library.2019.06.01.vec 
latin_library.lemmatized.2019.05.09.vec 
latin_library.non_lemmatized.2019.03.08.vec""".split()

[assess_word2vec(latin_words, embeddings_root, filename) for filename in vecs]

results:
assessing latin.lib.fasttext.vec
146 58415
....

**Comment by kylepjohnson** [comment]
OK thanks todd I'm doing better with gensim!

``` python
$ poetry run ipython
Python 3.7.5 (default, Oct 31 2019, 20:57:45) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.11.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from gensim.models.wrappers import FastText                             

In [2]: wiki_bin = "/Users/kyle.p.johnson/Downloads/wiki.la/wiki.la.bin"        

In [3]: wiki_vec = "/Users/kyle.p.johnson/Downloads/wiki.la/wiki.la.vec"        

In [4]: cc_bin = "/Users/kyle.p.johnson/Downloads/cc.la.300.bin"                

In [5]: cc_vec = "/Users/kyle.p.johnson/Downloads/cc.la.300.vec"                

In [6]: # for .bin: FastText.load_fasttext_format()                             

In [7]: # for .vec: FastText.load_word2vec_format()                             

In [8]: model = FastText.load_fasttext_format(wiki_bin)                         

In [9]: word = "amicitia"                                                       

In [10]: model.most_similar(word)                                               
Out[10]: 
[('amicitiam', 0.8625204563140869),
 ('amicitiae', 0.8571060299873352),
 ('amicitias', 0.8454258441925049),
 ('inimicitia', 0.7997336983680725),
 ('amicizia', 0.6959308385848999),
 ('pudicitia', 0.6888945698738098),
 ('inimicitiam', 0.6775547862052917),
 ('amici', 0.6744970083236694),
 ('inimicitiae', 0.6740451455116272),
 ('inimicitias', 0.6702579259872437)]

In [11]: model = FastText.load_fasttext_format(cc_bin)                          

In [12]: model.most_similar(word)                                               
Out[12]: 
[('amicitiá', 0.6483193635940552),
 ('amicitiâ', 0.6454339027404785),
 ('Inimicitia', 0.6269358992576599),
 ('amicitiaque', 0.6228713393211365),
 ('Amicitia', 0.6071096658706665),
 ('amicitias', 0.5946307182312012),
 ('amicitiae', 0.5937522649765015),
 ('inimicitia', 0.5720932483673096),
 ('amicitiam', 0.5685856342315674),
 ('amicitiarum', 0.5645438432693481)]

In [15]: from gensim.models import KeyedVectors                                 

In [16]: model = KeyedVectors.load_word2vec_format(wiki_vec)                    

In [17]: model.most_similar(word)                                               
Out[17]: 
[('amicitiam', 0.8625202178955078),
 ('amicitiae', 0.8571062088012695),
 ('amicitias', 0.845426082611084),
 ('inimicitia', 0.7997344136238098),
 ('amicizia', 0.6959313154220581),
 ('pudicitia', 0.6888941526412964),
 ('inimicitiam', 0.67755526304245),
 ('amici', 0.6744976043701172),
 ('inimicitiae', 0.6740456819534302),
 ('inimicitias', 0.6702578067779541)]

In [18]: model = KeyedVectors.load_word2vec_format(cc_vec)                      

In [19]: model.most_similar(word)                                               
Out[19]: 
[('amicitiâ', 0.6401441097259521),
 ('amicitiá', 0.6385208368301392),
 ('Inimicitia', 0.6270041465759277),
 ('amicitiaque', 0.6229627728462219),
 ('Amicitia', 0.607070803642273),
 ('amicitias', 0.5946215391159058),
 ('amicitiae', 0.593782365322113),
 ('inimicitia', 0.5720751881599426),
 ('amicitiam', 0.5685855150222778),
 ('amicitiarum', 0.5645272731781006)]
```

Interestingly, the newer models using Common Crawl data (plus Wiki) have presumably non-Latin  words: `amicitiâ', 'amicitiá'` @diyclassics Does one of these 4 look best to you? Per Todd's comment the .vec files might be a better long term move for us.

@free-variation You see any drawbacks to going solely with Gensim and avoiding the official `fasttext` library altogether?

**Comment by kylepjohnson** [comment]
One more, just the .vec from wiki and wiki+cc:

``` python
In [20]: word = "vocat"

In [22]: model = KeyedVectors.load_word2vec_format(wiki_vec)                    

In [23]: model.most_similar(word)                                               
Out[23]: 
[('lavocat', 0.7873127460479736),
 ('evocat', 0.7706935405731201),
 ('vocatur', 0.7136430740356445),
 ('advocat', 0.6994071006774902),
 ('invocat', 0.6955942511558533),
 ('revocat', 0.6897059082984924),
 ('vocans', 0.6792289614677429),
 ('vocati', 0.67292320728302),
 ('vocatos', 0.6660465002059937),
 ('provocat', 0.6538417339324951)]

In [24]: model = KeyedVectors.load_word2vec_format(cc_vec)                      

In [25]: model.most_similar(word)                                               
Out[25]: 
[('uocat', 0.574821949005127),
 ('prouocat', 0.5179796814918518),
 ('avocat', 0.4759005010128021),
 ('vocandam', 0.4686585068702698),
 ('Lavocat', 0.45595088601112366),
 ('πῖνον', 0.4528021514415741),
 ('advocat', 0.4495893120765686),
 ('voca', 0.44483232498168945),
 ('vocatos', 0.4362325668334961),
 ('vocas', 0.43096819519996643)]
```

@diyclassics how the heck does [πῖνον](http://www.perseus.tufts.edu/hopper/text?doc=Perseus%3Atext%3A1999.04.0057%3Aalphabetic+letter%3D*p%3Aentry+group%3D137%3Aentry%3Dpi%3Dnon) (?) end up in the Common Crawl data (that's been scraped from misc places online)? Between this and the accents above, I feel the wiki models will be safer.

**Comment by todd-cook** [comment]
πῖνον likely shows up because it's occurs in places similar to the query word; I've seen this happen somewhat rarely when rolling my own wordvec from a corpus. If the corpus is smallish, there won't be enough positive samples to drown out the example where a loanword is in close proximity. To get a clean monolingual embedding, one would have to drop loanwords from the source text.

**Comment by kylepjohnson** [comment]
> πῖνον likely shows up because it's occurs in places similar to the query word

Makes sense, though in that case I wonder about the data, as this is an obscure word (related to drinking) that I would not expect around vocat, even in the same languages. I wonder what % of Common Crawl data is thought to be Latin by (I assume) language identifiers.

**Comment by free-variation** [comment]
I think I may have lost the thread here :)  Are we still hoping to get fasttext to work, minimally with `.bin` files?

**Comment by kylepjohnson** [comment]
Closed, for now, with #39 . Will open some others to include auto DL of .bin files and allow non-FastText embeddings.
